### PR TITLE
feat(localization): add `lidar_marker_localizer`

### PR DESCRIPTION
**Pull request by SakodaShintaro**
## Description
Related discussion: https://github.com/orgs/autowarefoundation/discussions/4190

merge with https://github.com/autowarefoundation/autoware_launch/pull/861

## Tests performed

1. Download [sample dataset](https://drive.google.com/file/d/1FuGKbkWrvL_iKmtb45PO9SZl1vAaJFVG/view?usp=sharing)

_The sample dataset was acquired in National Institute for Land and Infrastructure Management, Full-scale tunnel experiment facility. The reflectors were installed by [Taisei Corporation](https://www.taisei.co.jp/english/)._

2. Launch Autoware with `pose_source:=ndt_lidar-marker`

```
ros2 launch autoware_launch logging_simulator.launch.xml \
  map_path:=$HOME/autoware_map/kokusouken \
  vehicle_model:=sample_vehicle \
  sensor_model:=sample_sensor_kit \
  pose_source:=ndt_lidar-marker
```
3. play rosbag

https://github.com/autowarefoundation/autoware.universe/assets/13819566/7eefb672-d5c9-4459-a525-b964582ffcbb

Please visualize the following topic in RViz.
- `/localization/pose_estimator/lidar_marker_localizer/pointcloud_preprocessor/ring_filter/pointcloud`
input lidar pointcloud
- `/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_mapped`
marker information listed in the vector map
- `/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_detected`
detected marker positions
- `/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_pointcloud`
detected marker pointcloud
- `/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/pose_with_covariance`
estimated position by lidar_marker_localizer

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by TaikiYamada4 — COMMENTED** [review]
@YamatoAndo 
I leaved some questions so could you answer them?
Plus, about outside of this PR, maybe you should right some explanation about this lidar_marker_localizer in [README.md of landmark_based_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/landmark_based_localizer/README.md)

File: localization/landmark_based_localizer/lidar_marker_localizer/flowchart.pu
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/flowchart.pu:1]: Thank you. I forgot about it.
I've added the flowchart to the readme. https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/ceaa67eb3747cf09b8445aa72dcbba617b4a1a13

File: launch/tier4_localization_launch/launch/pose_twist_estimator/pose_twist_estimator.launch.xml [hunk 1]
```diff
@@ -72,6 +73,16 @@
     </include>
   </group>
 
+  <!-- LiDAR Marker Localizer (as pose estimator) -->
+  <group if="$(var use_lidar_marker_pose)">
+    <push-ros-namespace namespace="pose_estimator"/>
+    <!-- TODO multi_localizer_mode -->
+    <include file="$(find-pkg-share tier4_localization_launch)/launch/pose_twist_estimator/lidar_marker_localizer.launch.xml">
+      <arg name="lidar_container_name" value="/sensing/lidar/top/pointcloud_preprocessor/pointcloud_container"/>
+      <arg name="input_pointcloud" value="/sensing/lidar/top/rectified/pointcloud_ex"/>
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply launch/tier4_localization_launch/launch/pose_twist_estimator/pose_twist_estimator.launch.xml:82]: Yes. /concatenated/pointcloud is not supported.

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+  points_sub_opt.callback_group = points_callback_group;
+  sub_points_ = this->create_subscription<PointCloud2>(
+    "~/input/pointcloud_ex", rclcpp::QoS(1).best_effort(),
+    std::bind(&LidarMarkerLocalizer::points_callback, this, _1), points_sub_opt);
+
+  rclcpp::CallbackGroup::SharedPtr self_pose_callback_group;
+  self_pose_callback_group =
+    this->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  auto self_pose_sub_opt = rclcpp::SubscriptionOptions();
+  points_sub_opt.callback_group = self_pose_callback_group;
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: It  should be `self_pose_sub_opt` :laughing: https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/ed76f70bcaced6ab79f2d54493e95ecf6ad9e767

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+  }
+  res->success = true;
+}
+
+std::vector<landmark_manager::Landmark> LidarMarkerLocalizer::detect_landmarks(
+  const PointCloud2::ConstSharedPtr & points_msg_ptr)
+{
+  const std::string sensor_frame = points_msg_ptr->header.frame_id;
+  // get sensor_frame pose to base_link
+  // TransformStamped transform_sensor_to_base_link;
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: I'll leave it as a TODO comment https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/6f234577a79106a7de511d7333d9c97481631218

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+  //   RCLCPP_WARN(get_logger(), "cannot get base_link to %s transform. %s", sensor_frame,
+  //   ex.what());
+  // }
+
+  pcl::PointCloud<autoware_point_types::PointXYZIRADRT>::Ptr points_ptr(
+    new pcl::PointCloud<autoware_point_types::PointXYZIRADRT>);
+  pcl::fromROSMsg(*points_msg_ptr, *points_ptr);
+
+  if (points_ptr->empty()) {
+    RCLCPP_WARN_STREAM_THROTTLE(this->get_logger(), *this->get_clock(), 1, "No points!");
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp:316]: Thank you! I thought the unit is seconds.
Then we'll need to fix the code for the ndt_scan_matcher as well.
https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp#L371

[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: Fixed https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/2d6de79a8093badfee6312f58d118585faf0c441

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+    distance_from_self_pose_to_nearest_marker <
+    param_.limit_distance_from_self_pose_to_nearest_marker;
+
+  if (!is_detected_marker) {
+    if (!is_exist_marker_within_self_pose) {
+      std::stringstream message;
+      message << "Could not detect marker, because the distance from self_pose to nearest_marker "
+                 "is too far ("
+              << distance_from_self_pose_to_nearest_marker << " [m]).";
+      // RCLCPP_INFO_STREAM_THROTTLE(this->get_logger(), *this->get_clock(), 1, message.str());
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: I've deleted this comment. https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/de39ba916c7f508620a15d61c32e36432fdcc552

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+      int64_t neg = 0;
+      double min_intensity = std::numeric_limits<double>::max();
+      double max_intensity = std::numeric_limits<double>::lowest();
+
+      // find max_min
+      for (size_t j = 0; j < param_.intensity_pattern.size(); j++) {
+        if (intensity_num[i + j] == 0) {
+          continue;
+        }
+        const double average = intensity_sum[i + j] / intensity_num[i + j];
```
[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: Good point! https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/e37e9f67b7bbca32bf9a9c333aa78f43b2f7a93b

**Comment by YamatoAndo** [comment]
@TaikiYamada4 
Thank you for your review.
I have completed all the corrections.

> Plus, about outside of this PR, maybe you should right some explanation about this lidar_marker_localizer in [README.md of landmark_based_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/landmark_based_localizer/README.md)

OK. I'll write something later

File: localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (TaikiYamada4) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: I think you need an exception when the intensity_num is 0 when you calculcate the average. Could you check this out?

[the remark this answers is not in the document]
Comment (YamatoAndo) [reply localization/landmark_based_localizer/lidar_marker_localizer/src/lidar_marker_localizer.cpp]: Fixed! https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/5565f67689a798163c86ecc61a57b35619f0c4a9

**Comment by TaikiYamada4** [comment]
@YamatoAndo 
Thank you for your revision!!
The content of this pull request seems good for me now, but the required checks (DCO and pre-commit.ci) haven't passed here, so could you check this out?

**Review by SakodaShintaro — COMMENTED** [review]
I think it's basically okay. I left a comment.

**Comment by SakodaShintaro** [comment]
Current status: Using the `/sensing/pointcloud_preprocessor/ring_filter/pointcloud_ex` topic as input is not consistent with the [interfaces defined in Autoware](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-architecture/sensing/data-types/point-cloud/) so it is not desirable.
Implementation is underway where the sensing component will output additional channel information, so we will wait for that to be implemented.

**Review by TaikiYamada4 — APPROVED** [review]
LGTM!

**Comment by SakodaShintaro** [comment]
[Current status] Waiting for the following PR

* https://github.com/autowarefoundation/autoware.universe/pull/6996

**Review by Motsu-san — APPROVED** [review]
Thank you for your pull-request.
I used this function on some tunnels and it was worked well.
[LGTMeow](https://lgtmeow.com)

**Comment by SakodaShintaro** [comment]
I checked this PR by TIER IV internal Evaluator.
https://evaluation.tier4.jp/evaluation/reports/cf23ec1f-f7c8-54e8-b9a2-5b549945d7b4?project_id=prd_jt

And I noticed that there is a issue with 
```
tier4_localization_launch: Cannot locate rosdep definition for [lidar_marker_localizer]
```

I will fix it.

**Comment by SakodaShintaro** [comment]
Sorry, I realized that the above problem has already been resolved at https://github.com/autowarefoundation/autoware.universe/pull/5573/commits/edabd49649d9534b9220476930288d60c2d50b2f

**Comment by TaikiYamada4** [comment]
@SakodaShintaro @YamatoAndo 
I see that the localization is going well, except I can't see the topic `/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_mapped` on rviz. What should it be like on rviz?

`ros2 topic echo` works and returns something like this.
```
header:
    stamp:
      sec: 1723008684
      nanosec: 791924782
    frame_id: map
  ns: landmark_text
  id: 18
  type: 9
  action: 0
  pose:
    position:
      x: 16518.486750000855
      y: 97361.17539999983
      z: 28.375
    orientation:
      x: 0.5453159159375941
      y: -0.45014503421135593
      z: -0.4501450342113559
      w: 0.5453159159375941
  scale:
    x: 0.0
    y: 0.0
    z: 0.5
  color:
    r: 1.0
    g: 0.0
    b: 0.0
    a: 1.0
  lifetime:
    sec: 0
    nanosec: 0
  frame_locked: false
  points: []
  colors: []
  texture_resource: ''
  texture:
    header:
      stamp:
        sec: 0
        nanosec: 0
      frame_id: ''
    format: ''
    data: []
  uv_coordinates: []
  text: (0)
  mesh_resource: ''
  mesh_file:
    filename: ''
    data: []
  mesh_use_embedded_materials: false
```

**Comment by YamatoAndo** [comment]
@TaikiYamada4 
Please set "transient local"

**Comment by TaikiYamada4** [comment]
@YamatoAndo 

Thanks! I found it, so the current PR here is approvable for me :+1:

**Comment by SakodaShintaro** [comment]
I was also able to confirm that it works too. This PR is ready to merge.

**Comment by cavadibrahimli1** [comment]
Screenshot from 2024-08-13 12-12-59
## Problem

I am encountering an issue where, despite successfully launching Autoware and loading the map, I am unable to see the car or point clouds in the RViz display. Additionally, this problem persists even after I attempt to initialize with GNSS.

## Commands Used

I used the following commands to launch the simulation and play the rosbag:

```bash
ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/autoware_map/kokusouken vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit pose_source:=ndt_lidar-marker
```

```bash
ros2 bag play autoware_map/kokusouken/rosbag/
```

## Steps to Reproduce

1. Launch Autoware with the provided commands.
2. Open RViz and observe that the car and point clouds are not visible.
3. Attempt to initialize with GNSS, but the issue remains unresolved.

## Expected Behavior

The car and point clouds should be visible in the RViz display, and GNSS initialization should allow proper visualization of the vehicle's position and sensor data.

## Actual Behavior

Neither the car nor the point clouds are visible in RViz, even after attempting GNSS initialization.

## Additional Information

- Running on UBUNTU 22.04 with ROS 2 humble 
- Attached a screenshot of the RViz display showing the current state.

**Comment by TaikiYamada4** [comment]
> Screenshot from 2024-08-13 12-12-59
> ## Problem
> 
> I am encountering an issue where, despite successfully launching Autoware and loading the map, I am unable to see the car or point clouds in the RViz display. Additionally, this problem persists even after I attempt to initialize with GNSS.
> ## Commands Used
> 
> I used the following commands to launch the simulation and play the rosbag:
> 
> ```shell
> ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/autoware_map/kokusouken vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit pose_source:=ndt_lidar-marker
> ```
> 
> ```shell
> ros2 bag play autoware_map/kokusouken/rosbag/
> ```
> 
> ## Steps to Reproduce
> 
>     1. Launch Autoware with the provided commands.
> 
>     2. Open RViz and observe that the car and point clouds are not visible.
> 
>     3. Attempt to initialize with GNSS, but the issue remains unresolved.
> 
> 
> ## Expected Behavior
> 
> The car and point clouds should be visible in the RViz display, and GNSS initialization should allow proper visualization of the vehicle's position and sensor data.
> ## Actual Behavior
> 
> Neither the car nor the point clouds are visible in RViz, even after attempting GNSS initialization.
> ## Additional Information
> 
>     * Running on UBUNTU 22.04 with ROS 2 humble
> 
>     * Attached a screenshot of the RViz display showing the current state.

@cavadibrahimli1 

The rosbag doesn't contain the `/clock` topic so you need the `--clock 200` option to your `ros2 bag play` command, but did you try that?

**Comment by cavadibrahimli1** [comment]
## Tests Performed

### 1. Start Autoware
```bash
ros2 launch autoware_launch logging_simulator.launch.xml \
  map_path:=$HOME/autoware_map/kokusouken \
  vehicle_model:=sample_vehicle \
  sensor_model:=sample_sensor_kit \
  pose_source:=ndt_lidar-marker
```

### 2. Correct Way of Running Rosbag
```bash
ros2 bag play autoware_map/kokusouken/rosbag/ --clock 200
```
@TaikiYamada4  

### 3. Play Rosbag Video
- A video was recorded to demonstrate the correct playback of the rosbag and successful visualization of the following topics in RViz:

```bash
/localization/pose_estimator/lidar_marker_localizer/pointcloud_preprocessor/ring_filter/pointcloud
/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_mapped
/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_detected
/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/marker_pointcloud
/localization/pose_estimator/lidar_marker_localizer/lidar_marker_localizer/debug/pose_with_covariance
```
[Screencast from 08-13-2024 01:26:21 PM.webm](https://github.com/user-attachments/assets/70e35a1c-9814-4045-976c-b5ca7858a49f)


## Effects on System Behavior

- **System Initialization**: The system initializes correctly when launching Autoware with the specified command, loading the map, vehicle model, and sensor configurations without issues.

- **Rosbag Playback**: Playing the rosbag with the `--clock 200` option generally ensures synchronized data processing, accurately reflecting real-time behavior. However, there are occasional instances where the rosbag deviates from the expected path, which may require manual intervention or resetting the playback to correct the trajectory.

- **Data Visualization**: The visualized topics in RViz, such as point clouds and marker positions, display as expected, confirming that the system’s localization and sensor data processing functions are operating correctly.

- **Performance Considerations**: During testing, the system performed well under normal conditions. However, it was observed that the system might freeze if a swap file is not created, particularly during memory-intensive operations. Creating a swap file is recommended to prevent these freezes and ensure smooth performance.

- **Resource Usage**: The system’s resource usage remains within acceptable limits under typical conditions. However, memory usage can spike during extensive data processing, emphasizing the need for adequate memory management, including the use of a swap file, to maintain system stability.
## Pre-review Checklist for the PR Author
```bash
[x] I've confirmed the contribution guidelines.
[x] The PR follows the pull request guidelines.
```

## In-review Checklist for the PR Reviewers
```bash
[ ] The PR follows the pull request guidelines.
```

## Post-review Checklist for the PR Author
```bash
[ ] There are no open discussions, or they are tracked via tickets.
```

**Comment by TaikiYamada4** [comment]
@cavadibrahimli1 
Is your comment an approval or something else?
It seems your video shows us the mismatch between the observed pointcloud and the pointcloud map.
Do you have any reasons why?

**Comment by TaikiYamada4** [comment]
@cavadibrahimli1 
This is a friendly reminder about my comment. What do you think about my previous comment?

**Comment by cavadibrahimli1** [comment]
@TaikiYamada4 Thank you for your follow-up. I apologize for not responding sooner—I got caught up with some other urgent tasks and unintentionally missed your earlier comment. 

I think the problem is with my computer. Because when I did the same test with my other computer, it worked without any problems. But in this test, there is a slight drift.

What do you think?

**Comment by TaikiYamada4** [comment]
> But in this test, there is a slight drift. What do you think?
It depends on what kind of drift is happening. Are you talking about the vehicle doesn't match expecially in the Z coordinates?
In your video it looks like the NDT is not working at all, but are you sure the ndt is outputting topics?

Plus, I want to ask your stance on this PR. Is your comment meant as an approval or a request change?

**Pull request merged**